src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/ClientConstructorAnalyzer.cs

```diff
@@ -45,14 +40,19 @@ public override void AnalyzeCore(ISymbolAnalysisContext context)
             {
                 if (constructor.DeclaredAccessibility == Accessibility.Public)
                 {
-                    if (IsClientOptionsParameter(constructor.Parameters.LastOrDefault()))
+                    var lastParameter = constructor.Parameters.LastOrDefault();
+
+                    if (IsClientOptionsParameter(lastParameter))
                     {
+                        // Allow optional options parameters
+                        if (lastParameter.IsOptional) continue;
+
                         var nonOptionsMethod = FindMethod(
-                            type.Constructors, constructor.Parameters.RemoveAt(constructor.Parameters.Length - 1));
+                            type.Constructors, constructor.TypeParameters, constructor.Parameters.RemoveAt(constructor.Parameters.Length - 1));
```

Comment:
Does it create a new array for every constructor?

Reply:
Yeah, thankfully there aren't that many.